tower-service-util/src/call_all/common.rs

```diff
@@ -0,0 +1,99 @@
+use super::Error;
+use futures::{try_ready, Async, Future, Poll, Stream};
+use tower_service::Service;
+
+/// TODO: Dox
+#[derive(Debug)]
+pub(crate) struct CallAll<Svc, S, Q> {
+    service: Svc,
+    stream: S,
+    queue: Q,
+    eof: bool,
+}
+
+pub(crate) trait Queue<T: Future> {
+    fn is_empty(&self) -> bool;
+
+    fn push(&mut self, future: T);
+
+    fn poll(&mut self) -> Poll<Option<T::Item>, T::Error>;
+}
+
+impl<Svc, S, Q> CallAll<Svc, S, Q>
+where
+    Svc: Service<S::Item>,
+    Svc::Error: Into<Error>,
+    S: Stream,
+    S::Error: Into<Error>,
+    Q: Queue<Svc::Future>,
+{
+    pub(crate) fn new(service: Svc, stream: S, queue: Q) -> CallAll<Svc, S, Q> {
+        CallAll {
+            service,
+            stream,
+            queue,
+            eof: false,
+        }
+    }
+
+    /// Extract the wrapped `Service`.
+    pub(crate) fn into_inner(self) -> Svc {
+        self.service
+    }
+
+    pub(crate) fn unordered(self) -> super::CallAllUnordered<Svc, S> {
+        assert!(self.queue.is_empty() && !self.eof);
+
+        super::CallAllUnordered::new(self.service, self.stream)
+    }
+}
+
+impl<Svc, S, Q> Stream for CallAll<Svc, S, Q>
+where
+    Svc: Service<S::Item>,
+    Svc::Error: Into<Error>,
+    S: Stream,
+    S::Error: Into<Error>,
+    Q: Queue<Svc::Future>,
+{
+    type Item = Svc::Response;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            let res = self.queue.poll().map_err(Into::into);
```

Comment:
I don't think you need the `map_err` here. `?` should automatically invoke `into`. And at that point, you can just put `self.queue.poll()?` directly in the `if let` below like in the original implementation.

Reply:
Unfortunately, you do :( Because `?` only works with `Box<Error>: From<E>` and not with `Into<Box<Error>>`

Reply:
Can't we just invert the bounds to match that instead then?
```rust
impl<Svc, S, Q> Stream for CallAll<Svc, S, Q>
where
    Error: From<Svc::Error>,
    Error: From<S::Error>,
    ...
```

Reply:
Then you can't use `String` / `&'static str` as errors.

Reply:
Ah, I guess this is pretty much https://github.com/rust-lang/rust/pull/58974 /cc @seanmonstar. I suppose we'll leave it this way for the time being.